MODAT-126: Implement OpenApi

steveellis opened the pull request:
Implements OpenApi for all non-filter routes. Removed code for things that OpenAPi takes care of like HTTP method validation and schema checks.

Separated non-filter and filter routing into their own classes.  Adding new non-filter routes only now needs modifications to the new RouteApi class.

Summary of the renaming:
- AuthorizeApi became FilterApi. It is exclusively devoted to filters now.
- AuthRoutingEntry becomes Route to simplify naming.

Changes to AuthTokenTest:
- Previously unsupported HTTP methods were 400, now they are 405. 
- We no longer check text in error responses that flow through the non-filter routes since OpenApi now generates those responses rather than our code.
- Renamed variables to better reflect the new token types instead of "basicToken" etc. Makes the code better but caused a large diff.
- Removed a duplicated variable (basicToken2).

Removed routes and tests for token encryption.

Tested on VM.

Added javadoc throughout.

Note: the `/token` route is the only route that is used in production. Refactored that and commented it to try to make it more clear that it handles POST data in two different shapes (for dummy signing requests and for access token signing requests). They only prop they share is `sub`.

julianladisch commented on src/main/java/org/folio/auth/authtokenmodule/apis/Api.java:
```diff
@@ -0,0 +1,47 @@
+    ctx.response().end(msg);
+  }
+
+  protected static void endJson(RoutingContext ctx, int code, String msg) {
```
Can you rename msg to json?

steveellis replied:
Sure no problem.

steveellis replied:
Done

julianladisch commented on src/test/java/org/folio/auth/authtokenmodule/AuthTokenTest.java:
```diff
@@ -131,14 +127,13 @@ public static void setUpClass(TestContext context) throws NoSuchAlgorithmExcepti
           } else {
+            var tenantAttributes = new JsonObject().put("module_to", "mod-authtoken-1.0.0");
+            initializeTenantForTokenStore(tenant, tenantAttributes);
             async.complete();
```
This should be futurized to avoid these nested if-clauses.

```
vertx.deployVerticle(...)
.compose(x -> vertx.deployVerticle(...))
.onComplete(context.asyncAssertSuccess(x -> {
  var tenantAttributes = new JsonObject().put("module_to", "mod-authtoken-1.0.0");
  initializeTenantForTokenStore(tenant, tenantAttributes);
}));
```

The logger.info is not needed.

steveellis replied:
Done

julianladisch commented on src/main/resources/headers/okapi-token.yaml:
```diff
@@ -0,0 +1,7 @@
+name: X-Okapi-Token
+description: Okapi Token
+# The /_/tenant route will not have this header.
+required: false
```
`required: false` means optional.
If some routes _require_ the token an okapi-token-required.yaml with `required: true` should be used.
If `/_/tenant` doesn't use a Token at all the okapi-token.yaml should be omitted.

steveellis replied:
Good catch. Yes agree.

steveellis replied:
See below for comment on how I propose we handle the dependency for `/_/tenant`.  But one side effect of this is that we'll need to leave this `required: false`. The other module that I'm aware of that uses folio-vertx-lib [also left it as](https://github.com/folio-org/mod-eusage-reports/blob/8cdf125331cd6b6fd31428d4b283062b4b1d7f1b/src/main/resources/openapi/headers/okapi-tenant.yaml#L4) `required: false` probably for this reason.

julianladisch commented on src/main/resources/headers/okapi-token.yaml:
```diff
@@ -0,0 +1,7 @@
+in: header
+name: X-Okapi-Token
+description: Okapi Token
+# The /_/tenant route will not have this header.
```
Which .yaml file contains the `/_/tenant` route?

steveellis replied:
Also good catch. No file does. I need to add it.

steveellis replied:
I think we may not want to replicate the yaml, but instead perhaps just note where to find the yaml for the /_/tenant route. It's here: https://github.com/folio-org/folio-vertx-lib/blob/master/core/src/main/resources/openapi/tenant-2.0.yaml. Duplicating may lead to bugs (not even sure if it would work). If we can make it clear in the comments what is going on (it's passing through the request to a handler in a dependent library folio-vertx-lib).

julianladisch commented on src/main/resources/openapi/token-1.0.yaml:
```diff
@@ -0,0 +1,110 @@
+openapi: 3.0.0
```
Is this ready to add a Jenkinsfile entry that lints and published the openapi files?
https://dev.folio.org/guides/jenkinsfile/#back-end-modules

steveellis replied:
Yes I think so. Will modify the Jenkins file.

steveellis replied:
Added. After moving the directories around the linting seems to be working. See what you think.

dcrossleyau replied:
Yes. I viewed the Artifacts apiLintReport. All good.

id-jenkins commented:
<html><body><pre>
INFO: api-lint: Using api-lint version: 1.1.0
INFO: api-lint: https://dev.folio.org/guides/api-lint/
INFO: api-lint: Assessing API description files: OAS
INFO: api-lint: Processing OAS 3.0 file: src/main/resources/openapi/token-1.0.yaml
ERROR: api-lint: 
---- Summary:
parsingResult.conforms: false
parsingResult.results.length: 35
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/refreshToken.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signRefreshToken.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/token.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signTokenPayload.json'
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: Unresolved reference 'schemas/signRefreshToken.json'
Violation: Unresolved reference 'schemas/signTokenPayload.json'
Violation: Unresolved reference 'schemas/refreshToken.json'
Violation: Unresolved reference 'schemas/token.json'
--------
validationResult.conforms: false
validationResult.results.length: 9
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
--------

---- parsingResult:
Model: file://src/main/resources/openapi/token-1.0.yaml
Conforms: false
Number of results: 35

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(8,14)-(8,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(9,14)-(9,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-url.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(10,14)-(10,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(34,14)-(34,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(35,14)-(35,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-url.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(36,14)-(36,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(59,14)-(59,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(60,14)-(60,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(61,8)-(62,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(61,14)-(61,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/refreshToken.json'
  Severity: Violation
  Target: schemas/refreshToken.json
  Property: 
  Range: [(104,12)-(104,37)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signRefreshToken.json'
  Severity: Violation
  Target: schemas/signRefreshToken.json
  Property: 
  Range: [(106,12)-(106,41)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/token.json'
  Severity: Violation
  Target: schemas/token.json
  Property: 
  Range: [(108,12)-(108,30)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signTokenPayload.json'
  Severity: Violation
  Target: schemas/signTokenPayload.json
  Property: 
  Range: [(110,12)-(110,41)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

---- validationResult:
ModelId: file://src/main/resources/openapi/token-1.0.yaml
Profile: OAS 3.0
Conforms: false
Number of results: 9

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(61,8)-(62,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml


ERROR: api-lint: There were processing errors. See list above.
</pre><body></html>


Note: When those ERRORs are fixed, then any INFO and WARNING messages will be gone too. See Jenkins "Artifacts" tab for ApiLintReport.

id-jenkins commented:
<html><body><pre>
INFO: api-lint: Using api-lint version: 1.1.0
INFO: api-lint: https://dev.folio.org/guides/api-lint/
INFO: api-lint: Assessing API description files: OAS
INFO: api-lint: Processing OAS 3.0 file: src/main/resources/openapi/token-1.0.yaml
ERROR: api-lint: 
---- Summary:
parsingResult.conforms: false
parsingResult.results.length: 35
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/refreshToken.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signRefreshToken.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/token.json'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signTokenPayload.json'
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-token.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
Violation: Unresolved reference 'schemas/signRefreshToken.json'
Violation: Unresolved reference 'schemas/signTokenPayload.json'
Violation: Unresolved reference 'schemas/refreshToken.json'
Violation: Unresolved reference 'schemas/token.json'
--------
validationResult.conforms: false
validationResult.results.length: 9
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
--------

---- parsingResult:
Model: file://src/main/resources/openapi/token-1.0.yaml
Conforms: false
Number of results: 35

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(8,14)-(8,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(9,14)-(9,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-url.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(10,14)-(10,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(34,14)-(34,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(35,14)-(35,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-url.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(36,14)-(36,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(59,14)-(59,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-token.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-token.yaml'
  Severity: Violation
  Target: headers/okapi-token.yaml
  Property: 
  Range: [(60,14)-(60,38)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_3
  Property: 
  Range: [(61,8)-(62,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(61,14)-(61,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/refreshToken.json'
  Severity: Violation
  Target: schemas/refreshToken.json
  Property: 
  Range: [(104,12)-(104,37)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signRefreshToken.json'
  Severity: Violation
  Target: schemas/signRefreshToken.json
  Property: 
  Range: [(106,12)-(106,41)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/token.json'
  Severity: Violation
  Target: schemas/token.json
  Property: 
  Range: [(108,12)-(108,30)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/schemas/signTokenPayload.json'
  Severity: Violation
  Target: schemas/signTokenPayload.json
  Property: 
  Range: [(110,12)-(110,41)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

---- validationResult:
ModelId: file://src/main/resources/openapi/token-1.0.yaml
Profile: OAS 3.0
Conforms: false
Number of results: 9

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(8,8)-(9,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(9,8)-(10,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Ftoken/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(10,8)-(11,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(34,8)-(35,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(35,8)-(36,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(36,8)-(37,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(59,8)-(60,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(60,8)-(61,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefreshtoken/parameter/parameter/default-binding/parameter_3/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(61,8)-(62,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml


ERROR: api-lint: There were processing errors. See list above.
</pre><body></html>


Note: When those ERRORs are fixed, then any INFO and WARNING messages will be gone too. See Jenkins "Artifacts" tab for ApiLintReport.

dcrossleyau commented:
The headers and schemas directories need to be inside the openapi directory.

steveellis commented:
> The headers and schemas directories need to be inside the openapi directory.

Thanks David! Was just puzzling over my sad jenkins report and see what you mean. :)

steveellis commented on src/main/resources/openapi/token-1.0.yaml:
```diff
@@ -0,0 +1,113 @@
+  # NOTE This module relies on the its folio-vertx-lib dependency to handle the /_/tenant
+  # route. Please see its yaml specification for /_/tenant here:
+  # https://github.com/folio-org/folio-vertx-lib/blob/master/core/src/main/resources/openapi/tenant-2.0.yaml
+  /token:
```
@julianladisch  see what you think. I'm thinking this comment is probably enough to address the handing of `/_/tenant`.

julianladisch replied:
I agree.
The official documentation link is https://dev.folio.org/reference/api/#folio-vertx-lib

steveellis replied:
Ok great, thanks Julian. And thanks for pointing out where the generated docs end up. I'll use that ink instead.

julianladisch commented on src/main/resources/openapi/headers/okapi-tenant.yaml:
```diff
@@ -0,0 +1,6 @@
+in: header
```
This overwrites the file from https://github.com/folio-org/folio-vertx-lib/tree/master/core/src/main/resources/openapi/headers
If the file is the same there is no need to overwrite it. If it is different a different file name should be used, for example okapi-tenant-required.yaml.
The same for okapi-token.yaml and okapi-url.yaml.

steveellis replied:
Ok thanks. I removed the yaml file for token (since isn't required and the underlying folio-vertx-yaml handles that case). I renamed the two files (url and tenant) to be `-required` since they change that underlying behavior.

steveellis commented on Jenkinsfile:
```diff
@@ -1,15 +1,21 @@
-      healthChk = 'no'
+      publishMaster = true
+      healthChk = true
+      healthChkCmd = 'curl -sS --fail -o /dev/null  http://localhost:8081/admin/health || exit 1'
```
Ok thanks. Changing to wget.

steveellis commented on src/main/java/org/folio/auth/authtokenmodule/apis/Api.java:
```diff
@@ -0,0 +1,47 @@
+      endText(ctx, e.getHttpResponseCode(), msg);
+      return;
+    }
+    logger.error("{}: {}", unexpectedExceptionMsg, h.toString());
```
I overrode `toString` in `TokenValidationException` [here](https://github.com/folio-org/mod-authtoken/blob/9ff15dc3f67343f742340527e51f4b2d06660156/src/main/java/org/folio/auth/authtokenmodule/tokens/TokenValidationException.java#L30) in order to provide the underlying cause to logging, but can see that if I call logger per your suggestion the cause is provided in the stack trace. So I removed the override. See what you think.

id-jenkins commented:
<html><body><pre>
INFO: api-lint: Using api-lint version: 1.1.0
INFO: api-lint: https://dev.folio.org/guides/api-lint/
INFO: api-lint: Assessing API description files: OAS
INFO: api-lint: Processing OAS 3.0 file: src/main/resources/openapi/token-1.0.yaml
ERROR: api-lint: 
---- Summary:
parsingResult.conforms: false
parsingResult.results.length: 6
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
Violation: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-tenant.yaml
Violation: 'name' property is required in a parameter.
Violation: Cannot find parameter or payload reference headers/okapi-url.yaml
--------
validationResult.conforms: false
validationResult.results.length: 2
Violation: 'in' property of a Parameter object is mandatory
Violation: 'in' property of a Parameter object is mandatory
--------

---- parsingResult:
Model: file://src/main/resources/openapi/token-1.0.yaml
Conforms: false
Number of results: 6

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(38,8)-(39,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-tenant.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1
  Property: 
  Range: [(38,8)-(39,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-tenant.yaml'
  Severity: Violation
  Target: headers/okapi-tenant.yaml
  Property: 
  Range: [(38,14)-(38,39)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#parameter-name-required
  Message: 'name' property is required in a parameter.
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(39,8)-(40,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#unresolved-parameter
  Message: Cannot find parameter or payload reference headers/okapi-url.yaml
  Severity: Violation
  Target: file://src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2
  Property: 
  Range: [(39,8)-(40,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/core#unresolved-reference
  Message: File Not Found: ENOENT: no such file or directory, open 'src/main/resources/openapi/headers/okapi-url.yaml'
  Severity: Violation
  Target: headers/okapi-url.yaml
  Property: 
  Range: [(39,14)-(39,36)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

---- validationResult:
ModelId: file://src/main/resources/openapi/token-1.0.yaml
Profile: OAS 3.0
Conforms: false
Number of results: 2

Level: Violation

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_1/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(38,8)-(39,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml

- Constraint: http://a.ml/vocabularies/amf/parser#Parameter-binding-minCount
  Message: 'in' property of a Parameter object is mandatory
  Severity: Violation
  Target: http://amferror.com/#errorParameter/file:/src/main/resources/openapi/token-1.0.yaml#/web-api/endpoint/%2Frefresh/parameter/parameter/default-binding/parameter_2/parameter/default-binding/default-name
  Property: http://a.ml/vocabularies/apiContract#binding
  Range: [(39,8)-(40,0)]
  Location: file://src/main/resources/openapi/token-1.0.yaml


ERROR: api-lint: There were processing errors. See list above.
</pre><body></html>


Note: When those ERRORs are fixed, then any INFO and WARNING messages will be gone too. See Jenkins "Artifacts" tab for ApiLintReport.

steveellis commented:
@julianladisch I mentioned this on slack but thought I would put more info here.

I'm thinking about rolling back [the last commit](https://github.com/folio-org/mod-authtoken/pull/122/commits/07599bc868ed1cffcc18f238941c14050660c2b6). In a [prior commit](https://github.com/folio-org/mod-authtoken/pull/122/commits/38af4c4305101c04183ce436c26de4b867e6646c) I removed the encryption endpoints. I think that is fine since they were in use and saved me having to implement them in OpenApi. We discussed that in our standup last week, but with this latest commit I removed encryption from the RT lifecycle.

I think removing this particular use of encryption might be a bit premature. We may end up wanting to encrypt RTs in storage for example, and if we do that we could also exchange them in an encrypted form as well. In any case we will be implementing the full RT lifecycle in a subsequent ticket it might be best to wait on this change until then.

From the [OAuth2 spec doc](https://datatracker.ietf.org/doc/html/rfc6749#section-10.4):

> Refresh tokens MUST be kept confidential in transit and storage.

There is more than one way of keeping tokens confidential in storage but encrypting the tokens would be one.

adamdickmeiss reviewed (approved):
Looks good and appears to work fine at least with my local setup.
